### PR TITLE
feat(workflow-executor): build database connection from split env vars

### PR DESCRIPTION
**Pull request by PMerlet**
## What

Let the workflow-executor CLI accept the Postgres connection either as a single `DATABASE_URL` **or** as split parts, for clients/platforms that expose database credentials as separate variables.

Resolution order:
- If `DATABASE_URL` is set → use it as-is (parts ignored).
- Otherwise → build the connection string from `DATABASE_HOST`, `DATABASE_NAME`, `DATABASE_USER` (required), `DATABASE_PASSWORD` (optional), `DATABASE_PORT` (optional, default `5432`).

If neither is provided (and not `--in-memory`), the startup error now lists both options. Partial parts raise a `ConfigurationError` naming exactly which parts are missing. User/password are URL-encoded so special characters are safe.

## Changes

- `src/cli-core.ts` — `resolveDatabaseUrl` / `buildDatabaseUrlFromParts`; updated missing-var message and `--help` text.
- `test/cli.test.ts` — new cases: build from parts, default port, no-password, URL-encoding, `DATABASE_URL` precedence, partial-parts error, non-numeric port, missing-vars message.
- `README.md` / `.env.example` — document the split variables.

## Testing

- `yarn workspace @forestadmin/workflow-executor lint` — clean (pre-existing warnings only)
- `yarn workspace @forestadmin/workflow-executor build` — passes
- `yarn workspace @forestadmin/workflow-executor test -- test/cli.test.ts` — 93 passed

🤖 Generated with [Claude Code](https://claude.com/claude-code)

> [!NOTE]
> ### Build database connection URL from split env vars in workflow-executor
> - Adds `buildDatabaseUrlFromParts` and `resolveDatabaseUrl` helpers in [cli-core.ts](https://github.com/ForestAdmin/agent-nodejs/pull/1727/files#diff-1605eabb559b21939e3b8d15b4a7a4c933da2a7cd5190022502db6ecaf93b396) to construct a Postgres URI from `DATABASE_HOST`, `DATABASE_NAME`, `DATABASE_USER`, `DATABASE_PASSWORD`, and `DATABASE_PORT` when `DATABASE_URL` is not set.
> - `DATABASE_URL` takes precedence when both are present; port defaults to 5432; password is optional; user/password are URL-encoded; IPv6 hosts are bracketed automatically.
> - `readEnvConfig` now validates that either `DATABASE_URL` or the three required parts (`DATABASE_HOST`, `DATABASE_NAME`, `DATABASE_USER`) are present, with updated error messaging; database vars are ignored in `--in-memory` mode.
> - Documents the two configuration approaches and precedence rules in [README.md](https://github.com/ForestAdmin/agent-nodejs/pull/1727/files#diff-84aa4115ae7aaddeae8ef0aa0f2f1fae55b5692e9eb4340a0ec43f2c3c3c0b3b) and [.env.example](https://github.com/ForestAdmin/agent-nodejs/pull/1727/files#diff-3a80e30ba98ef4aad58f8b208bc3a403357b99bbfd0d39a5affb7ea884c52501).
>
> 
> #### Changes since #1727 opened
>
> - Modified `buildDatabaseUrlFromParts` function in `workflow-executor` package to URL-encode database names when constructing postgres connection URIs from discrete environment variables [2ff18ad]
> - Added test coverage in `workflow-executor` package for database URL construction and environment variable configuration edge cases [2ff18ad]
> 
>
> 
>
> <sup><a href="https://app.macroscope.com">Macroscope</a> summarized ef63830.</sup>
> 
>
>

### CODE REVIEW

File: packages/workflow-executor/src/cli-core.ts [hunk 1]
```diff
@@ -169,12 +169,46 @@ function parseAiConfig(env: NodeJS.ProcessEnv): AiConfiguration[] | undefined {
...
+    );
+  }
+
+  const port = parsePositiveIntEnv('DATABASE_PORT', DATABASE_PORT) ?? DEFAULT_DATABASE_PORT;
+  const user = encodeURIComponent(DATABASE_USER as string);
+  const auth = DATABASE_PASSWORD ? `${user}:${encodeURIComponent(DATABASE_PASSWORD)}` : user;
+  const rawHost = DATABASE_HOST as string;
+  const host = rawHost.includes(':') && !rawHost.startsWith('[') ? `[${rawHost}]` : rawHost;
+
+  return `postgres://${auth}@${host}:${port}/${DATABASE_NAME}`;
```
Comment (hercemer42) [inline packages/workflow-executor/src/cli-core.ts]: **[Claude Opus 4.8 (1M context)] · Should fix** — `DATABASE_USER`/`DATABASE_PASSWORD` go through `encodeURIComponent` (L190-191) and the host is IPv6-bracketed (L193), but `DATABASE_NAME` is interpolated raw here. A name (or non-IPv6 host) containing a URL-reserved char (space, `/`, `?`, `#`, `@`) builds a malformed `postgres://` string that bypasses the up-front `ConfigurationError` and only fails opaquely at connect time. Such DB names are uncommon, so the real concern is the asymmetry — you already encode the auth segment and handle IPv6. Suggest `encodeURIComponent(DATABASE_NAME)` for consistency.

File: packages/workflow-executor/test/cli.test.ts [hunk 2]
```diff
@@ -279,6 +279,157 @@ describe('readEnvConfig', () => {
     expect(() => readEnvConfig({}, { ...args, inMemory: true })).toThrow(/FOREST_ENV_SECRET/);
   });
 
+  describe('database connection from parts', () => {
```
Comment (hercemer42) [inline packages/workflow-executor/test/cli.test.ts:282]: **[Claude Opus 4.8 (1M context)] · Preferential** — Strong, exact-value coverage on `readEnvConfig` here. The one gap: the parts→built-URI path is only asserted at the config level — no test drives a built-from-parts URI through `runCli` into `buildDatabase` (the `runCli` tests all use `baseEnv`'s literal `DATABASE_URL`). One `runCli([], { ...baseEnvNoUrl, DATABASE_HOST/NAME/USER }, factories)` asserting `buildDatabase` received `database: { uri: 'postgres://user@.../mydb' }` would prove the feature end-to-end, not just that the string is built.

File: packages/workflow-executor/src/cli-core.ts [hunk 3]
```diff
@@ -169,12 +169,46 @@ function parseAiConfig(env: NodeJS.ProcessEnv): AiConfiguration[] | undefined {
   ];
 }
 
+const DEFAULT_DATABASE_PORT = 5432;
+const DATABASE_PARTS = ['DATABASE_HOST', 'DATABASE_NAME', 'DATABASE_USER'] as const;
+
+function buildDatabaseUrlFromParts(env: NodeJS.ProcessEnv): string | undefined {
+  const { DATABASE_HOST, DATABASE_PORT, DATABASE_NAME, DATABASE_USER, DATABASE_PASSWORD } = env;
+
+  if (DATABASE_PARTS.every(key => !env[key])) return undefined;
```
Comment (hercemer42) [inline packages/workflow-executor/src/cli-core.ts:178]: **[Claude Opus 4.8 (1M context)] · Preferential** — This falsy check (and the `filter` on L180) treats a present-but-empty part as absent — load-bearing behavior with no test, whereas the same suite already tests empty-string `DATABASE_SSL`/`LOG_LEVEL`. Worth a case: all parts `''` → falls through to the missing-db-vars error (not a broken `postgres://@:5432/` URI); `DATABASE_HOST=''` with the other two set → throws the some-but-not-all error.

**Review by hercemer42 — CHANGES_REQUESTED** [review]

**Review by hercemer42 — APPROVED** [review]
